lib/core/src/Cardano/Wallet/Registry.hs

```diff
@@ -0,0 +1,199 @@
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies #-}
+
+module Cardano.Wallet.Registry
+    ( -- * Worker Registry
+      WorkerRegistry
+    , empty
+    , insert
+    , keys
+    , lookup
+    , remove
+
+      -- * Worker
+    , Worker
+    , MkWorker(..)
+    , newWorker
+    , workerThread
+    , workerId
+    , workerResource
+
+      -- * Context
+    , HasWorkerCtx (..)
+    ) where
+
+import Prelude hiding
+    ( log, lookup )
+
+import Cardano.BM.Trace
+    ( Trace, appendName )
+import Cardano.Wallet.Engine
+    ( HasLogger, logger )
+import Cardano.Wallet.Primitive.Types
+    ( WalletId )
+import Control.Concurrent
+    ( ThreadId, forkFinally, killThread )
+import Control.Concurrent.MVar
+    ( MVar
+    , modifyMVar_
+    , newEmptyMVar
+    , newMVar
+    , putMVar
+    , readMVar
+    , takeMVar
+    , tryPutMVar
+    )
+import Control.Exception
+    ( SomeException, finally )
+import Control.Monad.IO.Class
+    ( MonadIO, liftIO )
+import Data.Function
+    ( (&) )
+import Data.Generics.Internal.VL.Lens
+    ( (.~), (^.) )
+import Data.Generics.Labels
+    ()
+import Data.Generics.Product.Typed
+    ( HasType )
+import Data.Map.Strict
+    ( Map )
+import Data.Text
+    ( Text )
+import Data.Text.Class
+    ( toText )
+import GHC.Generics
+    ( Generic )
+
+import qualified Data.Map.Strict as Map
+import qualified Data.Text as T
+
+{-------------------------------------------------------------------------------
+                                Worker Context
+-------------------------------------------------------------------------------}
+
+-- | A class to link an existing context to a worker context.
+class HasType res (WorkerCtx ctx) => HasWorkerCtx res ctx where
+    type WorkerCtx ctx :: *
+    hoistResource :: res -> ctx -> WorkerCtx ctx
+
+{-------------------------------------------------------------------------------
+                                Worker Registry
+-------------------------------------------------------------------------------}
+
+-- | A registry to keep track of worker threads and wallet database connections.
+newtype WorkerRegistry res =
+    WorkerRegistry (MVar (Map WalletId (Worker res)))
+
+-- | Construct a new empty registry
+empty
+    :: IO (WorkerRegistry res)
+empty =
+    WorkerRegistry <$> newMVar mempty
+
+-- | Lookup the registry for a given worker
+lookup
+    :: MonadIO m
+    => WorkerRegistry res
+    -> WalletId
```

Comment:
Yeah, I thought the same. Didn't want to "bloat" this PR even more with this, but there's nothing particular about the 'WalletId' being a 'WalletId' here. So it could be anything in practice :+1: